synapse/rest/client/v2_alpha/forward_event.py

```diff
@@ -0,0 +1,134 @@
+from synapse.logging.opentracing import set_tag
+
+from ._base import client_patterns
+
+logger = logging.getLogger(__name__)
+
+
+class RoomEventForwardServlet(RestServlet):
+    """
+    PUT /net.maunium.msc2730/rooms/{room_id}/event/{event_id}/forward/{target_room_id}/{txn_id}
+    """
+
+    PATTERNS = client_patterns(
+        (
+            "/net.maunium.msc2730/rooms/(?P<room_id>[^/]*)/event/(?P<event_id>[^/]*)"
+            "/forward/(?P<target_room_id>[^/]*)/(?P<txn_id>.*)"
+        ),
+        releases=(),  # This is an unstable feature
+    )
+
+    _data_key = "net.maunium.msc2730.forwarded"
+    _err_not_forwardable = "NET.MAUNIUM.MSC2730_NOT_FORWARDABLE"
+
+    def __init__(self, hs):
+        super().__init__()
+        self.event_creation_handler = hs.get_event_creation_handler()
+        self.event_handler = hs.get_event_handler()
+        self.store = hs.get_datastore()
+        self.auth = hs.get_auth()
+
+    async def on_PUT(self, request, room_id, event_id, target_room_id, txn_id):
+        requester = await self.auth.get_user_by_req(request, allow_guest=True)
+
+        try:
+            event = await self.event_handler.get_event(
+                requester.user, room_id, event_id
+            )
+        except AuthError:
+            event = None
+        if not event:
+            raise SynapseError(404, "Event not found.", errcode=Codes.NOT_FOUND)
+
+        if event.is_state():
+            raise SynapseError(
+                401,
+                "State events cannot be forwarded.",
+                errcode=self._err_not_forwardable,
+            )
+        elif event.redacts:
+            raise SynapseError(
+                401,
+                "Redaction events cannot be forwarded.",
+                errcode=self._err_not_forwardable,
+            )
+        elif event.internal_metadata.is_redacted():
+            raise SynapseError(
+                401,
+                "Redacted events cannot be forwarded.",
+                errcode=self._err_not_forwardable,
+            )
+
+        event_id = event.event_id
+        event_dict = event.get_dict()
+
+        content = event_dict.pop("content")
+        unsigned = event_dict.pop("unsigned", {})
+        event_type = event_dict.pop("type")
+        has_forward_meta = self._data_key in content
+        try:
+            is_valid_forward = has_forward_meta and unsigned[self._data_key]["valid"]
+        except (KeyError, TypeError):
+            is_valid_forward = False
+
+        if has_forward_meta:
+            if not is_valid_forward:
+                raise SynapseError(
+                    401,
+                    "Event contains invalid forward metadata.",
+                    errcode=self._err_not_forwardable,
+                )
+            # Pass through the old event ID to the new unsigned data
+            event_id = unsigned[self._data_key]["event_id"]
+        elif not has_forward_meta:
+            content[self._data_key] = event_dict
```

Comment:
careful: I think this modifies the original event, stored in the cache. You need to copy `content` before modifying it.